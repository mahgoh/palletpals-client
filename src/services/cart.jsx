import { createContext, useContext, useState } from 'react'
import API from '@/services/api'

const CartContext = createContext()

export function useCart() {
  return useContext(CartContext)
}

export function CartProvider({ children }) {
  let [cart, setCart] = useState({
    totalCost: 0,
    shippingCost: 0,
    shoppingCart: [],
  })

  let refreshCart = async () => {
    try {
      let data = await API.Cart.get()
      setCart(data)
    } catch (e) {
      console.error(e)
    }
  }

  let value = { cart, setCart, refreshCart }
  return <CartContext.Provider value={value}>{children}</CartContext.Provider>
}
