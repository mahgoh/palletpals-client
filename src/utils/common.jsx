export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
