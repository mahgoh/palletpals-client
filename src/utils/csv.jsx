export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
