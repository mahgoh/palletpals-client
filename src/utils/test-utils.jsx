import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
