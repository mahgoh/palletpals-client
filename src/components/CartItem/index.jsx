import { NavLink } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { TrashIcon } from '@heroicons/react/outline'
import { useFormik } from 'formik'
import API from '@/services/api'
import { useCart } from '@/services/cart'
import { productImageURL } from '@/utils/common'
import Button from '@/components/Button'
import TextField from '@/components/TextField'

export default function CartItem({ item }) {
  const { t } = useTranslation()
  const { refreshCart } = useCart()

  if (!item) {
    return null
  }

  const { id, product, quantity, pricePerUnit } = item

  const validate = (values) => {
    const errors = {}
    if (!values.amount) {
      errors.amount = t('validation.required')
    } else if (values.amount < 1) {
      errors.amount = t('validation.min', { min: 1 })
    }

    return errors
  }

  const formik = useFormik({
    initialValues: {
      amount: quantity,
    },
    validate,
    onSubmit: async (values) => {
      try {
        await API.Cart.update(id, values.amount)
        await refreshCart()
        // TODO: Display success message
      } catch (e) {
        console.error(e)
      }
    },
  })

  async function removeItem() {
    try {
      await API.Cart.remove(id)
      await refreshCart()
      // TODO: Display success message
    } catch (e) {
      console.error(e)
    }
  }

  return (
    <div className="mb-4 flex flex-col last:mb-0 sm:flex-row">
      <div className="flex grow">
        <div className="h-32 w-32 overflow-hidden bg-gray-100 dark:bg-gray-800">
          {product.productImages.length > 0 && (
            <NavLink to={`/product/${product.id}`}>
              <img
                className="h-32 w-32 object-cover"
                src={productImageURL(product.productImages[0].id)}
                alt={product.name}
              />
            </NavLink>
          )}
        </div>
        <div className="flex flex-col py-2 px-6">
          <NavLink to={`/product/${product.id}`}>
            <h3 className="mb-2 text-lg font-bold hover:underline focus:underline">
              {product.name}
            </h3>
          </NavLink>
          <div>
            {t('common.price-per-unit')}: CHF {pricePerUnit.toFixed(2)}
          </div>
          <div>
            {t('common.amount')}: {quantity}
          </div>
          <div className="sm:mt-2">
            {t('common.total')}:{' '}
            <strong>CHF {(pricePerUnit * quantity).toFixed(2)}</strong>
          </div>
        </div>
      </div>
      <form
        className="mt-4 flex space-x-2 sm:mt-0 sm:flex-col sm:justify-end sm:space-x-0 sm:space-y-2"
        onSubmit={formik.handleSubmit}
      >
        <TextField
          type="number"
          placeholder={t('common.amount')}
          className="w-auto grow text-center sm:grow-0"
          showError={false}
          min={1}
          error={
            formik.touched.amount && formik.errors.amount
              ? formik.errors.amount
              : null
          }
          {...formik.getFieldProps('amount')}
        />
        <Button type="submit">{t('common.update')}</Button>
      </form>
      <div className="flex flex-col justify-end pt-2 sm:pt-0 sm:pl-4">
        <Button
          className="inline-flex justify-center"
          color="redOutline"
          title={t('common.remove')}
          onClick={removeItem}
        >
          <TrashIcon className="h-6 w-6" />
        </Button>
      </div>
    </div>
  )
}
