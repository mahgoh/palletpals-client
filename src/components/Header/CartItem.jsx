import { NavLink } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { productImageURL, formatPrice } from '@/utils/common'

export default function CartItem({ cartItem }) {
  const { t } = useTranslation()

  const { product, pricePerUnit, quantity } = cartItem

  return (
    <div className="flex py-4 first:pt-0 last:pb-0">
      <div className="mr-4 h-12 w-12 overflow-hidden bg-gray-100 dark:bg-gray-800">
        {product.productImages.length > 0 && (
          <NavLink to={`/product/${product.id}`}>
            <img
              className="h-12 w-12 object-cover"
              src={productImageURL(product.productImages[0].id)}
              alt={product.name}
            />
          </NavLink>
        )}
      </div>
      <div className="grow">
        <NavLink to={`/product/${product.id}`}>
          <h3 className="pb-2 text-lg font-bold leading-10 tracking-tight hover:underline focus:underline sm:text-lg sm:leading-none">
            {product.name}
          </h3>
        </NavLink>
        <div>
          {t('common.amount')}: {quantity}
        </div>
      </div>
      <div className="text-right text-lg font-medium">
        {formatPrice(pricePerUnit * quantity)}
      </div>
    </div>
  )
}
