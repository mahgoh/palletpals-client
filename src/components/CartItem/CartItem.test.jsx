import React from 'react'
import { render, screen } from '@/utils/test-utils'
import { cartItems } from '@/mocks/fixtures/cart'
import CartItem from '.'

describe('cart item', () => {
  it('should render', () => {
    const item = cartItems[0]
    const { container } = render(<CartItem item={item} />, {
      renderer: 'complete',
    })
    expect(screen.getByText(item.product.name)).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('should not render', () => {
    const { container } = render(<CartItem />, { renderer: 'complete' })
    expect(container).toBeEmptyDOMElement()
  })
})
