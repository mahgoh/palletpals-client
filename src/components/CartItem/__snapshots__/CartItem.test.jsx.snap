// Vitest Snapshot v1

exports[`cart item > should render 1`] = `
<div>
  <div
    class="mb-4 flex flex-col last:mb-0 sm:flex-row"
  >
    <div
      class="flex grow"
    >
      <div
        class="h-32 w-32 overflow-hidden bg-gray-100 dark:bg-gray-800"
      >
        <a
          class=""
          href="/product/4"
        >
          <img
            alt="Product A"
            class="h-32 w-32 object-cover"
            src="/api/product-images/2"
          />
        </a>
      </div>
      <div
        class="flex flex-col py-2 px-6"
      >
        <a
          class=""
          href="/product/4"
        >
          <h3
            class="mb-2 text-lg font-bold hover:underline focus:underline"
          >
            Product A
          </h3>
        </a>
        <div>
          Price per unit
          : 
          CHF 16.70
        </div>
        <div>
          Amount
          : 
          2
        </div>
        <div
          class="sm:mt-2"
        >
          Total
          :
           
          <strong>
            CHF 33.40
          </strong>
        </div>
      </div>
    </div>
    <form
      class="mt-4 flex space-x-2 sm:mt-0 sm:flex-col sm:justify-end sm:space-x-0 sm:space-y-2"
    >
      <div>
        <input
          class="w-auto grow text-center sm:grow-0 block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
          id="amount"
          min="1"
          name="amount"
          placeholder="Amount"
          type="number"
          value="2"
        />
      </div>
      <button
        class="bg-orange-500 focus:bg-orange-700 hover:bg-orange-700 focus:ring-orange-500 text-white px-6 py-2 sm:py-3 sm:px-8 items-center rounded-md border border-transparent text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
        type="submit"
      >
        Update
      </button>
    </form>
    <div
      class="flex flex-col justify-end pt-2 sm:pt-0 sm:pl-4"
    >
      <button
        class="inline-flex justify-center bg-rose-500/20 focus:bg-rose-500 hover:bg-rose-500 focus:ring-rose-500 text-rose-500 hover:text-white focus:text-white px-6 py-2 sm:py-3 sm:px-8 items-center rounded-md border border-transparent text-base font-medium ring-offset-transparent transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 "
        title="Remove"
      >
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
